### PR TITLE
#162685681 Like Comments

### PR DESCRIPTION
**Pull request by Opio-Emmanuel-Omona**
#### What does this PR do?
Authenticated users are able to like a specific comment

#### Description of Task to be completed?
Have the following endpoints working
```
PUT api/<slug>/comments/<int:id>/likes/        to like a comment/unlike a comment
```
```
GET api/<slug>/comments/<int:id>/likes/         to get all users that like a particular comment
```
```
GET api/<slug>/comments/<int:id>/likes/<pk>/        to get a specific like
```

#### How should this be manually tested?
Run the following commands
* git clone https://github.com/andela/Ah-backend-guardians.git
* cd Ah-backend-guardians
* python3 manage.py runserver

The open POSTMAN and test the endpoints

`for the PUT, no data is required.`

#### What are the relevant pivotal tracker stories?
#162685681

### CODE REVIEW

File: authors/apps/comments/renderers.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    charset = 'utf-8'
+
+    def render(self, data, media_type=None, renderer_context=None):
+
+        if 'errors' in data or 'detail' in data:
+            return super().render(data)
+
+        return json.dumps({
+            'likes': data,
+            # "likes count": data
```
Comment (moseskamira) [inline authors/apps/comments/renderers.py]: Hello, @Opio-Emmanuel-Omona  I have gone through your PR and realized you left some commented code. I humbly request you remove the comment.
Thank you.

Comment (Opio-Emmanuel-Omona) [reply authors/apps/comments/renderers.py]: @moseskamira I have replaced the commented out code with, a statement that counts the number of likes.

**Review by fahadmak — COMMENTED** [review]
Wouldn't it be more efficient to use one endpoint

File: authors/apps/comments/views.py [hunk 2]
```diff
@@ -125,3 +129,55 @@ def get(self, request, id, *args, **kwargs):
...
+    queryset = LikeComment.objects.all()
+
+    def destroy(self, request, *args, **kwargs):
+        like = self.get_object()
+        if request.user.pk == like.user.id:
+            self.perform_destroy(like)
+            return Response({"message": "You have unliked this comment"})
+        else:
+            raise ValidationError({
+                "message": "You do not like this comment"
```
Comment (fahadmak) [inline authors/apps/comments/views.py]: This error message is not descriptive enough. When a person doesn't have permission to delete a liked comment, THis message should tell him that he does not have permission to delete the liked comment.

**Review by fahadmak — CHANGES_REQUESTED** [review]

**Comment by Opio-Emmanuel-Omona** [comment]
> Wouldn't it be more efficient to use one endpoint

@fahadmak I've fixed the issue and now there's only one endpoint for liking and unliking a comment

File: authors/apps/comments/views.py (hunk 2 above)
Comment (Opio-Emmanuel-Omona) [reply authors/apps/comments/views.py]: @fahadmak The new by having one endpoint caters for this and therefore I have removed this entire validation.

**Review by tomuhenry — COMMENTED** [review]
Looks good... I see you worked on the like count we had talked about. And you're not switching between 2 different methods. Change the POST to PUT since the request body is empty then you're good to go

**Comment by Opio-Emmanuel-Omona** [comment]
> Looks good... I see you worked on the like count we had talked about. And you're not switching between 2 different methods. Change the POST to PUT since the request body is empty then you're good to go

@tomuhenry I have changed the POST to PUT

**Review by fahadmak — APPROVED** [review]

**Comment by fahadmak** [comment]
LGTM

**Review by tomuhenry — APPROVED** [review]
LGTM...

File: authors/apps/comments/views.py [hunk 3]
```diff
@@ -125,3 +128,49 @@ def get(self, request, id, *args, **kwargs):
...
+
+    def retrieve(self, request, *args, **kwargs):
+        likes = self.get_queryset()
+        serializer = self.serializer_class(data=likes, many=True)
+        serializer.is_valid()
+        return Response(serializer.data)
+
+    def update(self, request, *args, **kwargs):
+        id = self.kwargs.get('id')
+        comment = get_object_or_404(
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: @Opio-Emmanuel-Omona You should use the get_object method as suggested during the review session so as each method does what its supposed to do.

Comment (Opio-Emmanuel-Omona) [reply authors/apps/comments/views.py]: @archbishop12 I have included the get_object() method

File: authors/apps/comments/views.py [hunk 4]
```diff
@@ -125,3 +128,49 @@ def get(self, request, id, *args, **kwargs):
...
+        return get_object_or_404(Comment.objects.all(), id=id)
+
+    def retrieve(self, request, *args, **kwargs):
+        likes = self.get_queryset()
+        serializer = self.serializer_class(data=likes, many=True)
+        serializer.is_valid()
+        return Response(serializer.data)
+
+    def update(self, request, *args, **kwargs):
+        id = self.kwargs.get('id')
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: @Opio-Emmanuel-Omona remove this from the since you don't need it.

File: authors/apps/comments/views.py [hunk 5]
```diff
@@ -125,3 +128,49 @@ def get(self, request, id, *args, **kwargs):
...
+        comment = self.get_object()
+        like = {
+            'comment': comment.id,
+            'user': request.user.pk
+        }
+
+        if self.get_queryset().filter(comment=id, user=request.user.pk):
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: @Opio-Emmanuel-Omona You could do this to access the id.
```suggestion
        if self.get_queryset().filter(comment=comment.id, user=request.user.pk):
```

File: authors/apps/comments/views.py (hunk 4 above)
Comment (Opio-Emmanuel-Omona) [reply authors/apps/comments/views.py]: @archbishop12  I have removed it

File: authors/apps/comments/views.py (hunk 5 above)
Comment (Opio-Emmanuel-Omona) [reply authors/apps/comments/views.py]: @archbishop12  I am using what was suggested

File: authors/apps/comments/views.py [hunk 6]
```diff
@@ -125,3 +128,50 @@ def get(self, request, id, *args, **kwargs):
...
+        return Response(serializer.data)
+
+    def update(self, request, *args, **kwargs):
+        comment = self.get_object()
+        like = {
+            'comment': comment.id,
+            'user': request.user.pk
+        }
+
+        if self.get_queryset().filter(comment=comment.id,
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: This is repetition. Could you assign the whatever is returned by the filter to an object and then check that variable

```suggestion
       like = self.get_queryset().filter(comment=comment.id,
                                              user=request.user.pk)
        if like:
             like.delete()
             return ###
```

Comment (Opio-Emmanuel-Omona) [reply authors/apps/comments/views.py]: @archbishop12 I have removed the redundancy

**Review by archbishop12 — APPROVED** [review]
LGTM @Opio-Emmanuel-Omona I am merging this.

**Pull request merged**